compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeHashVisitor.java

```diff
@@ -563,19 +555,11 @@ private boolean isCyclic(BType type) {
         if (unresolvedTypes.contains(type)) {
             return true;
         }
-        visiting.push(type);
         unresolvedTypes.add(type);
         return false;
     }
 
     private Integer addToVisited(BType type, Integer hash) {
-        Integer existing = Optional.ofNullable(generated.get(hash)).orElse(0);
-
-        generated.put(hash, existing + 1);
-        if (existing > 0 && !visited.containsKey(type)) {
-            hash += existing;
-        }
-        assert visiting.pop() == type;
         visited.put(type, hash);
```

Comment:
should we have early return if type is already in visited?

Reply:
Early return is there before this method is called. Created https://github.com/ballerina-platform/ballerina-lang/issues/39689 to remove duplicate code.